llvm/lib/SYCLLowerIR/SanitizerKernelMetadata.cpp

```diff
@@ -38,6 +38,30 @@ PreservedAnalyses SanitizerKernelMetadataPass::run(Module &M,
   auto &DL = M.getDataLayout();
   auto &Ctx = M.getContext();
 
+  // Fix device global type, by wrapping a structure type
+  {
+    assert(KernelMetadata->getValueType()->isArrayTy());
+
+    auto Name = KernelMetadata->getName();
+    KernelMetadata->setName(Name + "_del");
+    auto *KernelMetadataOld = KernelMetadata;
+
+    StructType *StructTypeWithArray = StructType::create(Ctx);
+    StructTypeWithArray->setBody(KernelMetadataOld->getValueType());
+
+    KernelMetadata = new GlobalVariable(
+        M, StructTypeWithArray, false, GlobalValue::ExternalLinkage,
+        ConstantStruct::get(StructTypeWithArray,
+                            KernelMetadataOld->getInitializer()),
+        Name, nullptr, GlobalValue::NotThreadLocal, 1);
+    KernelMetadata->setUnnamedAddr(GlobalValue::UnnamedAddr::Local);
+    KernelMetadata->setDSOLocal(true);
+    KernelMetadata->copyAttributesFrom(KernelMetadataOld);
+    KernelMetadata->copyMetadata(KernelMetadataOld, 0);
+
+    KernelMetadataOld->eraseFromParent();
```

Comment:
why not generate KernelMetadataOld in struct type in the first place?

Reply:
This is a good question.

"KernelMetadata" is an array with appending linkage, so it can't be struct type at beginning.

I also considered to generate "KernelMetadata" at sycl-post-link step completely, but I use "KernelMetadata" to distinguish if the device image is a sanitized module at the same time. If we don't add "KernelMetadata" at sanitizer pass, it's possible that a sanitized module (but without any sanitized kernels) won't enable sanitizer layer at UR.

This is a design decision that if user uses "-fsanitize=" to compile program, we will always enable sanitizer layer in UR.